File: examples/tag_lookup_example.cpp
#include <cstdint>
#include <iostream>
#include <limits>
#include <string_view>

#include "dictionary_lookup.hpp"

namespace {

bool parse_tag_digits(char ch, std::uint32_t& value, int& digits) {
    unsigned digit = 0;
    if (ch >= '0' && ch <= '9') {
        digit = static_cast<unsigned>(ch - '0');
    } else if (ch >= 'A' && ch <= 'F') {
        digit = 10u + static_cast<unsigned>(ch - 'A');
    } else if (ch >= 'a' && ch <= 'f') {
        digit = 10u + static_cast<unsigned>(ch - 'a');
    } else {
        return false;
    }
    value = (value << 4) | digit;
    ++digits;
    return true;
}

bool parse_tag(std::string_view text, std::uint32_t& value_out) {
    std::uint32_t value = 0;
    int digits = 0;
    for (char ch : text) {
        if (ch == '(' || ch == ')' || ch == ',' || ch == ' ') {
            continue;
        }
        if (!parse_tag_digits(ch, value, digits)) {
            return false;
        }
    }
    if (digits != 8) {
        return false;
    }
    value_out = value;
    return true;
}

}  // namespace

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <tag> [<tag> ...]" << std::endl;
        std::cerr << "Accepts tags like (0010,0010) or 00100010" << std::endl;
        return 1;
    }

    for (int i = 1; i < argc; ++i) {
        const std::string_view text{argv[i]};
        std::uint32_t tag_value = 0;
        if (!parse_tag(text, tag_value)) {
            std::cout << text << ": invalid tag format" << std::endl;
            continue;
        }

        const auto* entry = dicom::lookup::tag_to_entry(tag_value);
        if (!entry) {
            std::cout << text << ": tag not found" << std::endl;
            continue;
        }

        std::cout << text << " -> keyword=" << entry->keyword << ", name=" << entry->name << std::endl;
    }

    return 0;
}

File: examples/python/tag_lookup_example.py
#!/usr/bin/env python3
"""Lookup DICOM tag metadata via dicomsdl."""

from __future__ import annotations

import argparse
import sys

import dicomsdl as dicom


def parse_tag(text: str) -> int:
    hex_digits = "".join(ch for ch in text if ch not in "() ,")
    if len(hex_digits) != 8:
        raise ValueError
    try:
        return int(hex_digits, 16)
    except ValueError as exc:  # pragma: no cover - same error channel
        raise ValueError from exc


def describe(entry: dict[str, object]) -> str:
    keyword = entry.get("keyword") or "<none>"
    retired = entry.get("retired") or ""
    retired_suffix = " (Retired)" if retired == "R" else ""
    return (
        f"keyword={keyword}, name={entry.get('name')}, VR={entry.get('vr')}"
        f"\n    tag={entry.get('tag')} vm={entry.get('vm')}{retired_suffix}"
    )


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(
        description="Lookup tag keyword/name from (gggg,eeee) or ggggeeee syntax",
    )
    parser.add_argument("tags", nargs="+", help="Tag strings like (0010,0010) or 00100010")
    args = parser.parse_args(argv)

    for text in args.tags:
        try:
            tag_value = parse_tag(text)
        except ValueError:
            print(f"{text}: invalid tag format")
            continue
        entry = dicom.tag_to_entry(tag_value)
        if entry is None:
            print(f"{text}: not found")
            continue
        print(f"{text} -> {describe(entry)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
